### PR TITLE
rpc: Implement `archive` RPC API

### PR DESCRIPTION
**Pull request by lexnv**
This PR implements the [archive](https://paritytech.github.io/json-rpc-interface-spec/api/archive.html) methods.

The `archive` methods include:
- [archive_unstable_body](https://paritytech.github.io/json-rpc-interface-spec/api/archive_unstable_body.html) - Get the extrinsics of a block
- [archive_unstable_genesisHash](https://paritytech.github.io/json-rpc-interface-spec/api/archive_unstable_genesisHash.html) - Get the genesis hash of the chain
- [archive_unstable_hashByHeight](https://paritytech.github.io/json-rpc-interface-spec/api/archive_unstable_hashByHeight.html) - Get the block hashes from a certain height of the block
- [archive_unstable_header](https://paritytech.github.io/json-rpc-interface-spec/api/archive_unstable_header.html) - Get the header of a block hash
- [archive_unstable_storage](https://paritytech.github.io/json-rpc-interface-spec/api/archive_unstable_storage.html) - Query storage of a block

This PR also includes the unsubscribe methods.

Part of: https://github.com/paritytech/substrate/issues/12071.

#### Manual Testing

```bash
archive_unstable_genesisHash: String("0x38c857c72005bf68bb1af4ac83405d94a9e2914850e33a4a3dfe8a5e5e2af8cd")
archive_unstable_hashByHeight on height=0 result: Ok(Object({"event": String("done"), "result": Array([String("0x38c857c72005bf68bb1af4ac83405d94a9e2914850e33a4a3dfe8a5e5e2af8cd")])}))
  archive_unstable_body hash="0x38c857c72005bf68bb1af4ac83405d94a9e2914850e33a4a3dfe8a5e5e2af8cd" result: Ok(Object({"event": String("done"), "result": String("0x00")}))
  archive_unstable_header hash="0x38c857c72005bf68bb1af4ac83405d94a9e2914850e33a4a3dfe8a5e5e2af8cd" result: Ok(Object({"event": String("done"), "result": String("0x000000000000000000000000000000000000000000000000000000000000000000483ec7c7e75d37442d10423ccef114efe2a4775125698d2d892113f85ee684c703170a2e7597b7b7e3d84c05391d139a62b157e78786d8c082f29dcf4c11131400")}))
archive_unstable_hashByHeight on height=1 result: Ok(Object({"event": String("done"), "result": Array([String("0xebcb7707bad09570beac92ca85abb47935574fe28515c9164e9a38df01ce2944")])}))
  archive_unstable_body hash="0xebcb7707bad09570beac92ca85abb47935574fe28515c9164e9a38df01ce2944" result: Ok(Object({"event": String("done"), "result": String("0x04280403000b98c6259b8401")}))
  archive_unstable_header hash="0xebcb7707bad09570beac92ca85abb47935574fe28515c9164e9a38df01ce2944" result: Ok(Object({"event": String("done"), "result": String("0x38c857c72005bf68bb1af4ac83405d94a9e2914850e33a4a3dfe8a5e5e2af8cd0418295f07a92d920bc1c281a5fee7531e9d0a71178e97a4747bc49be61677c528f7e7482f598e7e47123eda84423b8ee84328e023b6dd0209bc377c21f88d54e50c0642414245340200000000853a292100000000044241424529010104d43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d01000000000000000000000000000000000000000000000000000000000000000000000000000000054241424501014c890f007ffa8c3f0ecae68d41ec23a19ac9ede57c02e38c7266dbb4ec1e2d6e617dcb9ab96a80399b33939f7c4229308806c39eb6b34046c17731a90e7dae8a")}))
archive_unstable_hashByHeight on height=2 result: Ok(Object({"event": String("done"), "result": Array([String("0x106c5d1e1049f01e0036dd40c0c50599e55df4c552ea87c5d93b637c46ad71ba")])}))
  archive_unstable_body hash="0x106c5d1e1049f01e0036dd40c0c50599e55df4c552ea87c5d93b637c46ad71ba" result: Ok(Object({"event": String("done"), "result": String("0x04280403000b51d2259b8401")}))
  archive_unstable_header hash="0x106c5d1e1049f01e0036dd40c0c50599e55df4c552ea87c5d93b637c46ad71ba" result: Ok(Object({"event": String("done"), "result": String("0xebcb7707bad09570beac92ca85abb47935574fe28515c9164e9a38df01ce294408cd89427a3c56994eedaf1fc721339cd1bb8b363fa3ee937f1304a2c29c417119e99e9e8aee3b0effeb81f7fbd2fa3f5aa5d42ab953dc45fbe0fabeb04b60d077080642414245b5010100000000863a2921000000001ca2bd43a3b0d6a88b5a83f91ce93aa8f341650a87c262b524daf268a55b4738f9effe0bff6f33df69d679d9da706a820cb5adaa9c74ddd261d2d3b660db1d0ccf6f92023f3504ee99ceb3b3494235c242f90ea794535745de4313ae7e707304054241424501011a17a5890c212be368aeb28502ff9088c06896e5eac47953bcc7e1d016cc162bb375b169f924ecfe77db82816fe5ca0961da3346e9d080bbee4dceb44c038e8c")}))
archive_unstable_hashByHeight on height=3 result: Ok(Object({"event": String("done"), "result": Array([String("0x0b674621222e81a8db5b67386dffd6c24504c93d9080e6186149370226f7184a")])}))
  archive_unstable_body hash="0x0b674621222e81a8db5b67386dffd6c24504c93d9080e6186149370226f7184a" result: Ok(Object({"event": String("done"), "result": String("0x04280403000b08de259b8401")}))
  archive_unstable_header hash="0x0b674621222e81a8db5b67386dffd6c24504c93d9080e6186149370226f7184a" result: Ok(Object({"event": String("done"), "result": String("0x106c5d1e1049f01e0036dd40c0c50599e55df4c552ea87c5d93b637c46ad71ba0ca77e4418b7a1455bf8093fa79c6def40c595de82702d260b6a3507b0e45451d82e9e88b179af804e16c274917c4621e3deceea3e347a34fd5725df3407c03585080642414245b5010100000000873a2921000000002e99d1d4296672f4333e220c190666a08907925de0d10ed9c27a54789e20e76d896fb97d45a9c83d5b8bf760a3b93d3278b801f673663ffbd9fbb5782655990b8464470519e7d2ef9f7415ce75c67e12bc13c701e94981cb45d84b48e55dcb0d054241424501017c05029ba1a55026d41e70b755c08b345417b6e10875193a7447a52fc7659b19d5fcb9d09fa0990d4d6e42f157eecc6219b63dd19bb28e1c90b9c7f67601158b")}))

```

#### Unit Tests

`./src/archive/event.rs`

#### Integration Tests

 `./src/archive/tests.rs`

CC: @paritytech/tools-team.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-subkey
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2063027

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2063025

File: client/rpc-spec-v2/Cargo.toml [hunk 1]
```diff
@@ -20,14 +20,21 @@ sc-chain-spec = { version = "4.0.0-dev", path = "../chain-spec" }
...
 thiserror = "1.0"
 serde = "1.0"
+array-bytes = "4.1"
 hex = "0.4"
 futures = "0.3.21"
 
 [dev-dependencies]
 serde_json = "1.0"
-tokio = { version = "1.17.0", features = ["macros"] }
+tokio = { version = "1.17.0", features = ["macros", "full"] }
```
Comment (niklasad1) [inline client/rpc-spec-v2/Cargo.toml:35]: full implies macros :)

```suggestion
tokio = { version = "1.17.0", features = ["full"] }
```

File: client/rpc-spec-v2/src/archive/archive.rs [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+					// The block hash should have existed in the database. However,
+					// it may be possible that it was pruned.
+					vec![]
+				};
+
+				let _ = sink.send(&ArchiveEvent::Done(ArchiveResult { result }));
+				return
+			}
+
+			let finalized_hash = client.info().finalized_hash;
```
Comment (melekes) [inline client/rpc-spec-v2/src/archive/archive.rs:197]: what's the point of returning finalized blocks here if that's not what the client has asked for? shouldn't we return an error instead saying that "the block with the given height has not been finalized yet"?

File: client/service/src/builder.rs [hunk 3]
```diff
@@ -528,7 +528,7 @@ where
 			keystore.clone(),
 			system_rpc_tx.clone(),
 			&config,
-			backend.offchain_storage(),
+			backend.clone(),
```
Comment (melekes) [inline client/service/src/builder.rs:531]: so previously we're only accessing offchain storage and now we need full storage, right?

**Review by melekes — APPROVED** [review]
👍 (code-wise; don't know enough to reason about business logic)

File: client/rpc-spec-v2/src/archive/archive.rs (hunk 2 above)
Comment (lexnv) [reply client/rpc-spec-v2/src/archive/archive.rs:197]: Thanks for having a look at this!
I seem to have lost this in notifications, the `archive` class is not really necessary for light-client integrations and there are a few things to decide before moving forward with this:
https://github.com/paritytech/json-rpc-interface-spec/pull/26, https://github.com/paritytech/json-rpc-interface-spec/issues/27, https://github.com/paritytech/json-rpc-interface-spec/issues/30, https://github.com/paritytech/json-rpc-interface-spec/pull/29

Regarding this, we are also exploring adding another method `archive_finalizedBlock` to make sure users are aware they can make queries to ancestor blocks of finalized. IIRC, I placed this to make sure they are not making a call to any children of the finalized.

**Pull request closed without merging**